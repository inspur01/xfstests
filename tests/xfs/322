#! /bin/bash
# FS QA Test No. 322
#
# Reflink a file.
# Inject an error during rmap update to test log recovery.
#
#-----------------------------------------------------------------------
# Copyright (c) 2016, Oracle and/or its affiliates.  All Rights Reserved.
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License as
# published by the Free Software Foundation.
#
# This program is distributed in the hope that it would be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write the Free Software Foundation,
# Inc.,  51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
#-----------------------------------------------------------------------

seq=`basename $0`
seqres=$RESULT_DIR/$seq
echo "QA output created by $seq"

here=`pwd`
tmp=/tmp/$$
status=1    # failure is the default!
trap "_cleanup; exit \$status" 0 1 2 3 15

_cleanup()
{
	cd /
	_scratch_unmount > /dev/null 2>&1
	rm -rf $tmp.*
}

# get standard environment, filters and checks
. ./common/rc
. ./common/filter
. ./common/reflink
. ./common/inject

# real QA test starts here
_supported_os Linux
_supported_fs xfs
_require_cp_reflink
_require_scratch_reflink
_require_xfs_scratch_rmapbt
_require_xfs_io_error_injection "rmap_finish_one"

rm -f $seqres.full

blksz=65536
blks=64
sz=$((blksz * blks))
echo "Format filesystem"
_scratch_mkfs >/dev/null 2>&1
_scratch_mount >> $seqres.full

echo "Create files"
_pwrite_byte 0x66 0 $sz $SCRATCH_MNT/file1 >> $seqres.full
$XFS_IO_PROG -f -c "truncate $sz" $SCRATCH_MNT/file3 >> $seqres.full
sync

echo "Check files"
md5sum $SCRATCH_MNT/file1 | _filter_scratch
md5sum $SCRATCH_MNT/file3 | _filter_scratch

echo "Inject error"
_scratch_inject_error "rmap_finish_one"

echo "Try to reflink"
_reflink_range $SCRATCH_MNT/file1 0 $SCRATCH_MNT/file3 0 $sz >> $seqres.full

echo "FS should be shut down, touch will fail"
touch $SCRATCH_MNT/badfs 2>&1 | _filter_scratch

echo "Remount to replay log"
_scratch_inject_logprint >> $seqres.full

echo "Check files"
md5sum $SCRATCH_MNT/file1 | _filter_scratch
md5sum $SCRATCH_MNT/file3 | _filter_scratch

echo "FS should be online, touch should succeed"
touch $SCRATCH_MNT/goodfs

echo "Done"

# success, all done
status=0
exit
