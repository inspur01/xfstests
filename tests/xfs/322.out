QA output created by 322
Format filesystem
Create files
Check files
2a4f043bf9730a9e8882c9264b9797b3  SCRATCH_MNT/file1
b5cfa9d6c8febd618f91ac2843d50a1c  SCRATCH_MNT/file3
Inject error
Try to reflink
XFS_IOC_CLONE_RANGE: Input/output error
FS should be shut down, touch will fail
touch: cannot touch 'SCRATCH_MNT/badfs': Input/output error
Remount to replay log
Check files
2a4f043bf9730a9e8882c9264b9797b3  SCRATCH_MNT/file1
2a4f043bf9730a9e8882c9264b9797b3  SCRATCH_MNT/file3
FS should be online, touch should succeed
Done
